---
title: "STADyUM: Simulating and Analyzing Transcription Dynamics"
author: "Rebecca Hassett"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{STADyUM: Simulating and Analyzing Transcription Dynamics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

```{r load_package, include = FALSE}
# Load the package
library(STADyUM)
```

# Introduction

STADyUM is an R package for simulating and analyzing transcription dynamics.
It provides tools for:

- Simulating polymerase movement along genes
- Estimating transcription rates
- Analyzing pause site distributions
- Visualizing transcription dynamics
- Comparing transcription rates under different conditions

# Installation

```{r install, eval = FALSE}
# Install from GitHub
devtools::install_github("yourusername/STADyUM")
library(STADyUM)
```

# Basic Usage

## Estimating Transcription Rates from Experimental Data

First estimate transcription rates from an experiment's BigWig files containing
PRO-seq read counts without steric hindrance. Uses Expectation Maximization 
algorithm to estimate transcription rates such as gene body RNAP density (chi) 
and the ratio of gene body RNAP density to pause region RNAP density under
models of fixed pause sites (betaOrg) and variable pause sites (betaAdp).

The example data used throughout this package is derived from Aoi et al.
(2020). The data is from the raw PRO-Seq read counts for human cells
treated with auxin-induced acute depletion of NELF(negative elongation factor). 
Raw PRO-seq data is located at GEO: GSE144786. The raw fastq files were 
processed with https://github.com/Danko-Lab/proseq2.0. 

Estimates of transcription rates, such as $\Chi$ and $\Beta$, are calculated 
given Expectation Maximization with likelihood formulas derived in (Siepel 
2022, Section "Stationary Distribution and Inference at Steady-State") and 
stated in (Zhao 2023, Section "Inference at Steady State"). Note that $\Chi$ is 
the estimator for the average read depth, excluding the pause peak and $\Beta$ 
is the estimator for the ratio of $\Chi$ to the read depth in the pause peak 
region. The estimator for $\Beta$ is also given for a model adapted for varying
pause sites across cells and is denoted by variable beta_adp whereas the fixed 
pause site estimate is denoted by beta_org, refer to (Zhao 2023, Section 
"Allowing for variation in pause site"). Pause site statistics such as the mean
position of pause sites and the variance of pause sites (fkMean and fkVar) are 
also reported.


```{r estimate_experiment_rates}
load("../inst/extdata/granges_for_read_counting_DLD1_chr21.RData")

controlRates <- estimateTranscriptionRates(
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin_Ctrl-SE_plus_chr21.bw",
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin_Ctrl-SE_minus_chr21.bw",
bw_pause_filtered, bw_gb_filtered, "Control")

show(controlRates)

treatedRates <- estimateTranscriptionRates(
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin-SE_plus_chr21.bw", 
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin-SE_minus_chr21.bw", 
bw_pause_filtered, bw_gb_filtered, "Auxin Treated")

show(treatedRates)

controlRatesTbl <- rates(controlRates)

head(controlRatesTbl)
```

## Plotting functionality for TranscriptionRates objects

```{r plot_exp_rates, fig.width=5, fig.height=5}
plotMeanPauseDistrib(controlRates, file="mean_pause_site_distribution.png")

plotBetaVsChi(controlRates, betaType="betaAdp", file="controlBetaAdpVsChi.png")

plotExpectedVsActualPauseSiteCounts(treatedRates, 
                                    file="treatedExpActPauseSites.png")
```

## Estimate Transcription Rates for Experimental Data under Steric Hindrance

Under the model of steric hindrance, additional rates are estimated such as the 
landing pad occupancy (phi), pause-escape rate (betaZeta), the potential 
initiation rate (alphaZeta), and omegaZeta (effective initiation rate). The 
likelihood functions differ under the model of steric hindrance and are derived 
in (Zhao 2023, Section "Fitting the Steric Hindrance Model to the Data"). Note 
the omega parameter is a scaling factor.

```{r estimate_experiment_rates_steric_hindrance}
controlRatesSH <- estimateTranscriptionRates(
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin_Ctrl-SE_plus_chr21.bw",
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin_Ctrl-SE_minus_chr21.bw",
bw_pause_filtered, bw_gb_filtered, "Control", stericHindrance=TRUE, 
omegaScale=12.3768278981277)

show(controlRatesSH)

treatedRatesSH <- estimateTranscriptionRates(
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin-SE_plus_chr21.bw", 
"../inst/extdata/PROseq-DLD1-aoi-NELFC_Auxin-SE_minus_chr21.bw", 
bw_pause_filtered, bw_gb_filtered, "Auxin Treated", stericHindrance=TRUE, 
omegaScale=11.0318379571379)

show(treatedRatesSH)
```


## Likelihood Ratio Tests

Compare samples treated with auxin to control samples using the 
likelihood ratio test functionality derived in (Siepel 2022, Section 
"LikelihoodRatio Tests for Differences Between Transcription Units, Conditions, 
or Species")

```{r likelihood_ratio_tests}
lrt <- likelihoodRatioTest(controlRates, treatedRates)

show(lrt)
```

## Plot Results from Likelihood Ratio Tests

As expected given reports from (??), RNAPs tend to pause more downstream when 
NELF is absent due to Auxin treatment. As shown in the plot, the majority of 
treated genes have average pause locations of more than 100bp downstream of the 
TSS. Strikingly, few RNAPs could enter productive elongation after NELF 
depletion and treated samples had a signficantly lower pause-escape rate than 
the control samples exemplified by the $\Beta$ violin plots.

```{r plot_lrt_results, fig.width=5, fig.height=5}

plotPauseSiteContourMapTwoConditions(lrt, file="lrtPauseSiteContourMap.png")

plotMeanPauseDistrib(controlRates, file="meanPauseSitesControl.png")
plotMeanPauseDistrib(treatedRates, file="meanPauseSitesTreated.png")

BetaViolinPlot(lrt, file="lrtBetaViolinPlot.png")

ChiViolinPlot(lrt, file="lrtChiViolinPlot.png")

plotLfcMa(lrt, file="lrtMAplot.png")
```

## Simulate Polymerase

![(A) Conceptual illustration of model focusing on the kinetic model for RNAP 
movement on the DNA template. (B) Graphical model representation with unobserved
continuous-time Markov chain (Z) and observed read counts (X). Read counts at 
each site are conditionally independent and Poisson distributed. (C) Design of 
SimPol simulator tracking the movement of in silico RNAPs across N-bp DNA 
templates in C cells, then samples the synthetic read counts based on RNAP 
positions. (D) Example of synthetic nascent RNA sequencing data from SimPol 
alongside matched real PRO-seq data from the DNAJA1 gene on chromosome 9 of the 
human genome](simpol_figures.png)

Simulate read counts by running simulations of polymerase movement along the
gene at user-specified rates and gene lengths. Our simulator models the movement
of RNAPs along a set of transcription units given a continuous-time Markov 
model. Nascent RNA sequencing read counts are then generated under steady-state
conditions by sampling based on a Poisson distribution with mean equal to the 
simulated RNAP frequency multiplied by a user-specified sequencing depth. A 
graphical model representation (Figure B) with unobserved continuous-time Markov
chain (Z_i) in which user-specified paramters for transcription initiation 
(alpha), pause-escape, and elongation (zeta) are included as well as the 
observed read counts (Xi). The initial model is extended to allow for 
variability across cells in promoter-proximal pause site locations and steric 
hindrance of transcription initiation from paused RNAPs (Figure C). More 
information on the simulator can be found in (Zhao 2023, Sections "A 
simple probabilistic model for transcription initiation, promoter-proximal 
pausing, and elongation", "SimPol Simulator", "Generation of synthetic NRS 
data")

```{r create_simpol, fig.width=10, fig.height=8}
simpol <- simulatePolymerase(k=50, ksd=25, kMin=17, kMax=200, geneLen=1950,
alpha=2, beta=0.5, zeta=2000, zetaSd=1000, zetaMin=1500, zetaMax=2500,
cellNum=1000, polSize=33, addSpace=17, time=40, timesToRecord=c(10, 20, 30))

show(simpol)

readCounts <- readCounts(simpol)

plotPauseSites(simpol, file="simpolPauseSites.png")
plotPositionHeatmap(simpol, file="polymerasePositionHeatmap.png")
plotCombinedCells(simpol, file="combinedCellsLollipopPlot.png")
```

## Estimate Transcription Rates from Simualted Data

```{r estimate_simulated_rates, fig.width=10, fig.height=8}
simRates <- estimateTranscriptionRates(simpol, name="simRates1a2b0.5", stericHindrance=TRUE)

show(simRates)

plotChiDistrib(simRates, file="simChis.png")
```


## Session Info
```{r session_info}
sessionInfo()
```

# References
- Yuki Aoi, Edwin R. Smith, Avani P. Shah, Emily J. Rendleman, Stacy A. 
Marshall, Ashley R. Woodfin, Fei X. Chen, Ramin Shiekhattar, Ali Shilatifard,
NELF Regulates a Promoter-Proximal Step Distinct from RNA Pol II Pause-Release,
Molecular Cell,Volume 78, Issue 2,2020,Pages 261-274.e5,ISSN 1097-2765,
https://doi.org/10.1016/j.molcel.2020.02.014. 
- Yixin Zhao, Lingjie Liu, Rebecca Hassett, Adam Siepel, Model-based 
characterization of the equilibrium dynamics of transcription initiation and 
promoter-proximal pausing in human cells, Nucleic Acids Research, Volume 51, 
Issue 21, 27 November 2023, Page e106, https://doi.org/10.1093/nar/gkad843
- Chu, T., Wang, Z., Chou, S. P., & Danko, C. G. (2018). Discovering 
Transcriptional Regulatory Elements From Run‐On and Sequencing Data Using the 
Web‐Based dREG Gateway. Current protocols in bioinformatics, e70.
- Lingjie Liu, Yixin Zhao, Rebecca Hassett, Shushan Toneyan, Peter K Koo, Adam 
Siepel, Probabilistic and machine-learning methods for predicting local rates 
of transcription elongation from nascent RNA sequencing data, Nucleic Acids
Research, Volume 53, Issue 4, 28 February 2025, gkaf092, 
https://doi.org/10.1093/nar/gkaf092
- Adam Siepel, A Unified Probabilistic Modeling Framework for Eukaryotic 
Transcription Based on Nascent RNA Sequencing Data, bioRxiv, 1, January, 2022